sound/soc/sof/core.c

```diff
@@ -50,6 +49,37 @@ static const struct sof_panic_msg panic_msg[] = {
 	{SOF_IPC_PANIC_ASSERT, "assertion failed"},
 };
 
+static void sof_client_dev_register(struct snd_sof_dev *sdev, const char *name)
+{
+	struct snd_sof_client *client;
+
+	client = devm_kzalloc(sdev->dev, sizeof(*client), GFP_KERNEL);
+	if (!client)
+		goto err;
+
+	/* register client platform device */
+	client->pdev = platform_device_register_data(sdev->dev, name,
+						     PLATFORM_DEVID_NONE,
+						     client, sizeof(*client));
+	if (IS_ERR(client->pdev))
+		goto err;
+
+	list_add(&client->list, &sdev->client_list);
```

Comment:
Does this need locking ? i.e. are we using this in IRQ handler/thread or in other context sensitive places ?

Reply:
hmm at the moment, the only place where this is called is by the SOF core at the end of its probe callback.

Reply:
What about the IRQ thread ? i.e when we are propagating through the clients when host receives an IPC message (when determining which client gets the IPC).

Reply:
@lgirdwood this brings the other question in my mind, are clients allowed to register themselves with the SOF core? Or are the clients supposed to be registered by the SOF core itself as has been done for the 2 clients here?

Reply:
SOF core exports the registration APIs, its up to clients to then register when they probe().  The mediator will also be a client, clients can also be bespoke.

Reply:
@lgirdwood makes sense but if I expose an API to register, how would I get a handle to the SOF core device in this API to be able to add the client to the client list?

Reply:
We need something like device registration on Linux e.g.
```
/* description of client to help with registration and routing of IPCs */
struct sof_driver {
    const char *name;
    enum sof_drv_class class; // audio, sensing,  etc
    int id;
    // any ops ?
};

struct sof_client * snd_soc_client_register(struct sof_drive *drv);
```

Reply:
@lgirdwood Sorry maybe I am missing something basic here. In the current implementation, the assumption is that all client devices are children of the SOF core device and they are created and registered by the core itself. But your ask goes a bit further is requiring any device to be able to register with the SOF core. So my question is how does this client device get a pointer to the SOF core device in order to be able to register itself?

Reply:
Clients dont initially need a pointer, see above. Clients pass in a "driver" structure and are returned a pointer. This way any client can register without have to modify core.

Reply:
This does mean that there has to be a static that holds the pointer to the SOF core device that the register function has access to, right?

Reply:
Yes, it's the same as generic device/driver registration.

Reply:
@lgirdwood I remember Pierre mentioned (a couple of years ago) that static variables are generally not recommended in kernel drivers. But this is inevitable if we want clients to register themselves with the SOF core. @plbossart do you have an opinion here?

Reply:
why would this be a static? You can devm-allocate a context during the probe of the client module, and keep the sof_client handle there.

Reply:
@plbossart we're talking about the pointer to SOF core device, sdev, in the above line. When a client register with the SOF core by invoking the client register API, we need add it the list of the client in sdev. But if sdev is not static, how else can we access it?

Reply:
As I mentioned it, you guys are overloading the notion of 'register'.

In the code above, what you are doing is create a child device. Of course the client cannot create such a device on its own, it can only provide a driver for such a child without controlling when the child device is actually created.

If what you want is have the client request a communication channel with the core, then it's a different registration mechanism different from the usual device/driver handling.

Reply:
I might add that I am confused by @lgirdwood statements. In the case where the 'client' will be handling an audio card, so essentially be running a machine driver, then how exactly would this client probe on its own? We have zero mechanisms in ACPI for this at the moment.

Reply:
> As I mentioned it, you guys are overloading the notion of 'register'.
> 
> In the code above, what you are doing is create a child device. Of course the client cannot create such a device on its own, it can only provide a driver for such a child without controlling when the child device is actually created.
> 
> If what you want is have the client request a communication channel with the core, then it's a different registration mechanism different from the usual device/driver handling.

@plbossart this is the part I'm struggling with in Liam's request. What would that alternate registration mechanism be?

Reply:
@plbossart @ranj063 I'm asking you guys to copy the same flows as the Linux driver model. There will probably need to be a ```static LIST_HEAD(client_drivers)``` somewhere in the SOF core (please look at device core, last time I check many years ago it did the same).

1) We register client drivers NOT client devices, the client devices are then ```probed()``` as part of the enumeration. ASoC does this too with components, but there would be no binding or similar enumeration dependencies needed for SOF clients (except that base FW has booted). 
2) We must always pass ```struct sof_client_dev``` in all our APIs. We must not use ```struct device``` as the handle and abuse the private_data between client and core. We can also back cast ```sof_client_dev``` too for any sub APIs that are invented later.

Should I create some example code ?

Reply:
Liam, can you please elaborate on "We register client drivers NOT client devices, the client devices are then probed() as part of the enumeration."

In the case of machine drivers, how exactly would this work? The machine driver is handled as a platform driver, so someone needs to create the device for the machine driver probe to be executed.

Reply:
@lgirdwood thanks for the clarification. I think I get the idea now except for the part about who registers the client drivers. In the case of ASoC Components, the platform driver registers the component drv. But who would do this in the case of SOF clients?

Reply:
I don't think we should have any dependency now on ASoC machine driver probing in order to probe and boot the DSP (this may not be the case today), this way non audio functions do not depend on audio. Can someone confirm if this is the case (sorry, I last deep dived the code about a year ago) ? 

I also think we have to consider two registration uses cases (especially if we move to IPC2.0 with discoverability).

1) Base FW contains five features, all which need a feature driver. SOF core then creates the platform devices for these features on FW boot enumeration data. i.e. FW boot data contains support for audio class, hence SOF infrastructure creates audio platform device (audio driver registration is done by audio class driver).

1.1) SOF core infrastructure probes and boots FW based on PCI/ACPI/DT ID.
1.2) SOF base FW boot meta data contains all supported features and is parsed by SOF core
1.3) SOF core creates platform devices (using existing machine driver names and lookup tables based on machine) for each function.

2) Base FW contains limited functions. Users need to load functions on ad-hoc basis, the sof client driver registers (as part of module insmod flow) and FW module (pointed to by driver structure) is loaded at this point too. This then causes base FW to load the module and emit module boot data (which contains class info) causing the SOF infrastructure to create platform device.

3) rmmod case for 2) above. base FW unloads module, SOF infrastructure destroys platform device and SOD client driver is unregistered.

Does this make sense ? My main thinking here is that we can no longer call the current code the SOF driver but rather the SOF infrastructure.

Reply:
@lgirdwood I was ok until the last paragraph. If we create a new type of 'sof_driver', then it means we need to have an sof_device, and given the recent complexities of doing such things with SoundWire we may want to stick to platform devices for now?

Reply:
@plbossart it's just a container for platform_driver, you can then add our own ops and stuff in there too, likewise sof_client can be a container of platform_device (again we add our own stuff around it - rather than breaking the platform device platform data for clients).

Reply:
that could work technically indeed if we extend platform drivers, but it's still a pain since we'd modify legacy machine drivers. I am not sure I want to touch all this.

well that said if we take out HDMI from those machine drivers we have to take the scissors anyways. Gah, more work...

Reply:
> I don't think we should have any dependency now on ASoC machine driver probing in order to probe and boot the DSP (this may not be the case today), this way non audio functions do not depend on audio. Can someone confirm if this is the case (sorry, I last deep dived the code about a year ago) ?

@lgirdwood There's no dependency on machine drv probing today in order to boot the DSP but there is a dependency on finding an ACPI match. But that has all been removed in this PR